DIPs/DIP1xxx-bottom_type.md

```diff
@@ -0,0 +1,206 @@
+#Add Bottom Type
+
+##Abstract
+
+Introduce a bottom type which is a type that has no values.
+
+The primary purpose is to specify
+the return type of a function that does not return.
+
+
+##Links
+
+[Bottom type](https://en.wikipedia.org/wiki/Bottom_type)
+
+
+##Terminology
+
+For the purpose of this document, the bottom type is referred to
+as *Tbottom*, aka \U+22A5, &uptack;.
+
+
+##Rationale
+
+Being able to specify that a function does not return has
+advantages:
+
+1. Smaller/faster code - no stack cleanup code has to be emitted
+after such calls, no registers used on that branch need to be
+preserved, variables not used on that branch will not be marked
+as 'live', variables reassigned on that branch will not affect
+other branches.
+2. Makes it easier to reason about the code by simplifying control flow.
+3. It's necessary to be competitive with other systems programming
+languages.
+4. RAII cleanup, scope, finally, and catch clauses need not be generated for
+such branches.
+
+
+##Description
+
+Since assert(0) is defined to never return,
+it is used to generate a bottom type:
+
+```
+alias Tbottom = typeof(assert(0));
+```
+
+and is used like:
+
+```
+Tbottom fatalError(string message) { ... }
+```
+
+Properties of Tbottom:
+
+```
+Tbottom.mangleof == "??";
+Tbottom.sizeof == 0;
+Tbottom.alignsize == 1;
+```
+
+Type construction:
+
+```
+Tbottom -> Tbottom
+const(Tbottom) -> Tbottom
+immutable(Tbottom) -> Tbottom
+shared(Tbottom) -> Tbottom
+inout(Tbottom) -> Tbottom
+Tbottom* -> Tbottom
+Tbottom[] -> Tbottom
+Tbottom[integer] -> Tbottom
+T[Tbottom] -> Tbottom
+Tbottom[T] -> Tbottom
+(Tbottom) -> (Tbottom)
+```
+
+A Tbottom is implicitly convertible to any other type.
+No other type is implicitly convertible to Tbottom.
+
+##Declarations
+
+A variable, parameter, field or enum may not be declared as
+Tbottom.
+
+A function that returns auto may infer the return type as Tbottom.
+(Breaking Change - used to be void.)
```

Comment:
This is mutually exclusive with the claim made in the `##Breaking Changes / Deprecation Process` section.

Reply:
I added that to the Breaking Changes section in my edits before I initiated the community review.